### PR TITLE
Add Validate() to all collections

### PR DESCRIPTION
**Pull request by Doridian**
This also fixes a bug in `internal/reflect/map.go` where it used the map's own type to Validate elements, which does not seem correct at all.

Fixes https://github.com/hashicorp/terraform-plugin-framework/issues/467

### CODE REVIEW

File: types/set.go [hunk 1]
```diff
@@ -148,15 +148,23 @@ func (st SetType) Validate(ctx context.Context, in tftypes.Value, path path.Path
 		return diags
 	}
 
+	validatableType, isValidatable := st.ElemType.(xattr.TypeWithValidate)
```
Comment (bflad) [inline types/set.go:151]: We should ensure there is a covering unit test for this change. 👍

File: types/map.go [hunk 2]
```diff
@@ -115,6 +117,54 @@ func (m MapType) String() string {
...
+// Validate validates all elements of the map that are of type
+// xattr.TypeWithValidate.
+func (m MapType) Validate(ctx context.Context, in tftypes.Value, path path.Path) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if in.Type() == nil {
+		return diags
+	}
+
+	if !in.Type().Is(tftypes.Map{}) {
```
Comment (bflad) [inline types/map.go:129]: We should ensure there is a covering unit test for this new conditional. 👍

**Review by bflad — CHANGES_REQUESTED** [review]
Hi @Doridian 👋 Thank you for submitting this. Overall this is looking pretty good, it could just use some additional unit testing and small adjustments. If you have any questions about adding that testing, please reach out.

Also, if you could, it would be great to include a changelog entry which can be done by adding a `.changelog/481.txt` file with contents such as:

``````markdown
```release-note:bug
types: Ensured `List`, `Map`, and `Set` types with `xattr.TypeWithValidate` elements run validation on those elements
```
``````

**Comment by Doridian** [comment]
It seems that `AtSetValue` expects an `attr.Value` which is not present in `Validate`. Pushed a fix for that.

**Comment by bflad** [comment]
@Doridian since we are adding a method onto the type, we can unit test the method itself, e.g.

```go
func TestListTypeValidate(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		listType      ListType
		tfValue       tftypes.Value
		path          path.Path
		expectedDiags diag.Diagnostics
	}{
		"wrong-value-type": {
			listType: ListType{
				ElemType: StringType,
			},
			tfValue: tftypes.NewValue(tftypes.Set{
				ElementType: tftypes.String,
			}, []tftypes.Value{
				tftypes.NewValue(tftypes.String, "testvalue"),
			}),
			path: path.Root("test"),
			expectedDiags: diag.Diagnostics{
				diag.NewAttributeErrorDiagnostic(
					path.Root("test"),
					"List Type Validation Error",
					"An unexpected error was encountered trying to validate an attribute value. This is always an error in the provider. Please report the following to the provider developer:\n\n"+
						"expected List value, received tftypes.Value with value: tftypes.Set[tftypes.String]<tftypes.String<\"testvalue\">>",
				),
			},
		},
		"no-validation": {
			listType: ListType{
				ElemType: StringType,
			},
			tfValue: tftypes.NewValue(tftypes.List{
				ElementType: tftypes.String,
			}, []tftypes.Value{
				tftypes.NewValue(tftypes.String, "testvalue"),
			}),
			path: path.Root("test"),
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			diags := testCase.listType.Validate(context.Background(), testCase.tfValue, testCase.path)

			if diff := cmp.Diff(diags, testCase.expectedDiags); diff != "" {
				t.Errorf("unexpected diagnostics difference: %s", diff)
			}
		})
	}
}
```

Rinse and repeat for `MapType` and `SetType`.

This is okay for now as it would _unfortunately_ take a lot more effort to import the existing `internal/testing/types.StringTypeWithValidateError` to verify the element validation diagnostics are passed back, as importing that right now would cause an import cycle and updating the testing file to `package types_test` and putting `types.` in front of everything in the `types` package is a lot to ask. 😅

**Comment by bflad** [comment]
(We could also create an unexported type with validation errors in the test files too, but it is totally fine.)

**Comment by Doridian** [comment]
I have added tests for list and map exactly as you described.
For map, as it already was validatable, I have just appended the `wrong-value-type` test case.

**Comment by Doridian** [comment]
> This is okay for now as it would _unfortunately_ take a lot more effort to import the existing `internal/testing/types.StringTypeWithValidateError` to verify the element validation diagnostics are passed back, as importing that right now would cause an import cycle and updating the testing file to `package types_test` and putting `types.` in front of everything in the `types` package is a lot to ask. 😅

Maybe in another PR if I find the time 😉

**Review by bflad — APPROVED** [review]
This looks great, @Doridian, thank you so much for your efforts here 🚀

**Pull request merged**